code/hpke/Hacl.Impl.HPKE.fst

```diff
@@ -0,0 +1,492 @@
+     as_seq h0 pskID_hash));
+  let infolen_b = sub output (7ul +. nsize_dh_public cs +. nsize_dh_public cs +. nsize_dh_public cs +. 1ul +. nhash_length cs) 1ul in
+  Lib.ByteBuffer.uint_to_bytes_be infolen_b (nhash_length_u8 cs);
+  (**) let h7 = get() in
+  (**) Lib.ByteSequence.lemma_uint_to_bytes_be_preserves_value (nhash_length_u8 cs);
+  (**) Lib.ByteSequence.nat_from_intseq_be_inj (as_seq h7 infolen_b) (Lib.ByteSequence.nat_to_bytes_be #SEC 1 hlength);
+  (**) assert (as_seq h7 infolen_b `Seq.equal` Lib.ByteSequence.nat_to_bytes_be #SEC 1 hlength);
+  (**) assert (as_seq h7 (gsub output 0ul (7ul +. nsize_dh_public cs +. nsize_dh_public cs +. nsize_dh_public cs +. 2ul +. nhash_length cs)) `Seq.equal`
+    (S.id_of_mode S.Base `Seq.append`
+    S.id_of_cs cs `Seq.append`
+    as_seq h0 pkE `Seq.append`
+    as_seq h0 pkR `Seq.append`
+    as_seq h0 pkI `Seq.append`
+    Lib.ByteSequence.nat_to_bytes_be 1 hlength `Seq.append`
+    as_seq h0 pskID_hash `Seq.append`
+    Lib.ByteSequence.nat_to_bytes_be 1 hlength));
+  let output_info = sub output (7ul +. nsize_dh_public cs +. nsize_dh_public cs +. nsize_dh_public cs +. 2ul +. nhash_length cs) (nhash_length cs) in
+  (**) assert(disjoint output_info info_hash);
+  copy output_info info_hash;
+  (**) let h8 = get() in
+  (**) assert (as_seq h8 (gsub output 0ul (7ul +. nsize_dh_public cs +. nsize_dh_public cs +. nsize_dh_public cs +. 2ul +. nhash_length cs +. nhash_length cs)) `Seq.equal`
+    (S.id_of_mode S.Base `Seq.append`
+    S.id_of_cs cs `Seq.append`
+    as_seq h0 pkE `Seq.append`
+    as_seq h0 pkR `Seq.append`
+    as_seq h0 pkI `Seq.append`
+    Lib.ByteSequence.nat_to_bytes_be 1 hlength `Seq.append`
+    as_seq h0 pskID_hash `Seq.append`
+    Lib.ByteSequence.nat_to_bytes_be 1 hlength `Seq.append`
+    as_seq h0 info_hash))
+
+noextract inline_for_extraction
+val ks_derive_default_aux:
+     #cs:S.ciphersuite
+  -> pkR:key_dh_public cs
+  -> zz:key_dh_public cs
+  -> pkE:key_dh_public cs
+  -> infolen: size_t{v infolen <= S.max_info}
+  -> info: lbuffer uint8 infolen
+  -> o_key: key_aead cs
+  -> o_nonce: nonce_aead cs
+  -> context_len:size_t{v context_len == 9 + (3 * S.size_dh_public cs) + 2 *
+                                     Spec.Agile.Hash.size_hash (S.hash_of_cs cs)}
+  -> context:lbuffer uint8 context_len
+  -> secret:lbuffer uint8 (nhash_length cs)
+  -> pkI:lbuffer uint8 (nsize_dh_public cs)
+  -> psk:lbuffer uint8 (nhash_length cs)
+  -> label_key:lbuffer uint8 8ul
+  -> label_nonce:lbuffer uint8 10ul
+  -> tmp:lbuffer uint8 (10ul +. context_len)
+  -> ST unit
+       (requires fun h0 ->
+         live h0 pkR /\ live h0 zz /\ live h0 pkE /\
+         live h0 info /\ live h0 o_key /\ live h0 o_nonce /\
+         live h0 context /\ live h0 secret /\ live h0 pkI /\
+         live h0 psk /\ live h0 tmp /\ live h0 label_key /\ live h0 label_nonce /\
+         disjoint o_key o_nonce /\ disjoint secret psk /\ disjoint secret zz /\
```

Comment:
any chance you could use an `all_disjoint` predicate here for readability?

Reply:
Unfortunately, since I'm using hacl's buffers and not F*'s buffers, the two interact badly together, there is a type mismatch between lists. I'd be happy to use bigops predicate if there is an easy way to solve this, but I think the root cause is https://github.com/FStarLang/FStar/issues/65

Reply:
can you use FStar.BigOps to define all_disjoint and all_live in Lib.Buffer? might as well?

Reply:
I could try, but my guess would be that we would have the same problem with buffer vs lbuffer (I already had type issues in this code when I had a lbuffer while expecting a buffer for instance, and that's without using it inside a list constructor)

Reply:
Following discussion, I used `LowStar.Monotonic.Buffer.all_disjoint` for readability here. `all_live` seems trickier, but also less important since liveness predicates are not pairwise.